### PR TITLE
*: only add default value for final aggregation to fix the aggregate push down (partition) union case

### LINKED ISSUE #35295: static pruning mode gets wrong results
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE `t100` (
`ID` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
`col1` int(10) NOT NULL DEFAULT '0' COMMENT 'test',
`money` bigint(20) NOT NULL COMMENT 'test',
`logtime` datetime NOT NULL COMMENT '记录时间',
PRIMARY KEY (`ID`,`logtime`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin AUTO_INCREMENT=1 COMMENT='test'
PARTITION BY RANGE COLUMNS(logtime) (
PARTITION `p20220608` VALUES LESS THAN ("20220609"),
PARTITION `p20220609` VALUES LESS THAN ("20220610"),
PARTITION `p20220610` VALUES LESS THAN ("20220611"),
PARTITION `p20220611` VALUES LESS THAN ("20220612"),
PARTITION `p20220612` VALUES LESS THAN ("20220613"),
PARTITION `p20220613` VALUES LESS THAN ("20220614"),
PARTITION `p20220614` VALUES LESS THAN ("20220615"),
PARTITION `p20220615` VALUES LESS THAN ("20220616"),
PARTITION `p20220616` VALUES LESS THAN ("20220617"),
PARTITION `p20220617` VALUES LESS THAN ("20220618"),
PARTITION `p20220618` VALUES LESS THAN ("20220619"),
PARTITION `p20220619` VALUES LESS THAN ("20220620"),
PARTITION `p20220620` VALUES LESS THAN ("20220621"),
PARTITION `p20220621` VALUES LESS THAN ("20220622"),
PARTITION `p20220622` VALUES LESS THAN ("20220623"),
PARTITION `p20220623` VALUES LESS THAN ("20220624"),
PARTITION `p20220624` VALUES LESS THAN ("20220625")
 );

insert into t100(col1,money,logtime) values (100,10,'2022-06-09 00:00:00');

insert into t100(col1,money,logtime) values (100,10,'2022-06-10 00:00:00');

SELECT col1,sum(money) FROM t100 WHERE logtime>='2022-06-09 00:00:00' AND col1=100 ;
```

### 2. What did you expect to see? (Required)
```
mysql> SELECT col1,sum(money) FROM t100 WHERE logtime>='2022-06-09 00:00:00' AND col1=100 ;
+------+------------+
| col1 | sum(money) |
+------+------------+
|  100 |         20 |
+------+------------+
```

### 3. What did you see instead (Required)
```
mysql> SELECT col1,sum(money) FROM t100 WHERE logtime>='2022-06-09 00:00:00' AND col1=100 ;
+------+------------+
| col1 | sum(money) |
+------+------------+
| NULL |         20 |
+------+------------+
```
Sometimes will get this result. Try more times.

### 4. What is your TiDB version? (Required)
5d4745e9d41532e8b7a5bf5cc3df894bafacaa00

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #35295

Problem Summary:

When table `t` is empty, this get empty result:

```
select avg(id) from t group by id;
```

While this get `NULL`:

```
select avg(id) from t;
```

We add `NULL` (default value) to the empty result here: https://github.com/pingcap/tidb/blob/2de01b4ee2e774da47f08edc11df469dba524f05/executor/aggregate.go#L1368-L1379

Whether or not add the `NULL` row depends on some conditions.

Note, for this case, the added `NULL` row make the final result wrong:

image

### What is changed and how it works?

Only add `NULL` for final aggregation.
Otherwise the add `NULL` row operation would be done several times and send a wrong `NULL` row to the final aggregation.

- In aggregation push down optimze rule, I'll set the pushed down aggregation to non-final
- In physical plan, aggregation push down to cop phase, I'll not blindly change the upper aggregation to final/complete... because it's parent might be final and it's a partial itself 
- In the executor builder phase, only set the default value for final aggregation

### Check List

Tests 

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @XuHuaiyu @qw4990

File: expression/aggregation/descriptor.go [hunk 1]
```diff
@@ -123,8 +123,6 @@ func (a *AggFuncDesc) Split(ordinal []int) (partialAggDesc, finalAggDesc *AggFun
 		partialAggDesc.Mode = Partial1Mode
 	} else if a.Mode == FinalMode {
 		partialAggDesc.Mode = Partial2Mode
-	} else {
```
Comment (XuHuaiyu) [inline expression/aggregation/descriptor.go:126]: Why remove this check?

File: planner/core/physical_plans.go [hunk 2]
```diff
@@ -1077,6 +1077,10 @@ type basePhysicalAgg struct {
 	MppPartitionCols []*property.MPPPartitionColumn
 }
 
+func (p *basePhysicalAgg) IsFinalAgg() bool {
```
Comment (XuHuaiyu) [inline planner/core/physical_plans.go:1080]: Why not just rename the old func?

File: planner/core/task.go [hunk 3]
```diff
@@ -1417,7 +1421,11 @@ func BuildFinalModeAggregation(
 				}
 			}
 
-			finalAggFunc.Mode = aggregation.FinalMode
+			if aggFunc.Mode == aggregation.CompleteMode {
```
Comment (XuHuaiyu) [inline planner/core/task.go:1424]: I do not understand what does this code block mean

File: expression/aggregation/descriptor.go (hunk 1 above)
Comment (tiancaiamao) [reply expression/aggregation/descriptor.go:126]: In the previous assumption, there will be no partial mode agg before coprocessor push down...

But now I change the logical plan aggregation to partial, so in the coprocessor push down here, this line may meet partial mode agg.

File: planner/core/task.go (hunk 3 above)
Comment (tiancaiamao) [reply planner/core/task.go:1424]: In the old assumption, **before the coprocessor push down, the agg mode is either Complete or Final**. So after the push down, the parent agg become Final mode.

But in the new assumption, before the coprocessor push down, the agg mode can be partial. So here the parent agg mode is set to different value accordingly.

For final before, we set it to final.
For partial, it's set to partial2

File: planner/core/physical_plans.go (hunk 2 above)
Comment (tiancaiamao) [reply planner/core/physical_plans.go:1080]: Done

File: planner/core/task.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply planner/core/task.go:1424]: Agg1->PartitionUnion->Agg2->TableReader

In coprocessor push down, Agg2->TableReader become Agg3(root)->TableReader->Agg4(cop)->TableScan(cop)
In the past, Agg2 is always Complete or Final, but in this PR here Agg2 could be Partial
So `BuildFinalModeAggregation`  need to consider the difference here.

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply planner/core/task.go:1432]: I add some comment in the code so the reviewer know what happen.

**Review by winoros — APPROVED** [review]

File: planner/core/rule_aggregation_push_down.go [hunk 4]
```diff
@@ -438,6 +438,16 @@ func (a *aggregationPushDownSolver) tryAggPushDownForUnion(union *LogicalUnionAl
 	if pushedAgg == nil {
 		return nil
 	}
+
+	// Update the agg mode for the pushed down aggregation.
```
Comment (bb7133) [inline planner/core/rule_aggregation_push_down.go:442]: Could you explain why this update happens?

I don't get the logic to change 'CompleteMode' -> 'Partial1Mode'...

Comment (tiancaiamao) [reply planner/core/rule_aggregation_push_down.go:442]: **The behaviour between partition / final is different.**
Take `count(id)` for example:

```
a XX                  a 1                         a 1
b YY                  b 1                         b 2
c ZZ                  c 1                         c 2
     (partial mode)==>         (final mode) ==>  

b SS                  b 1
c KK                  c 1
```





In partial mode, the input is original data and turned into   "xx 1" where xx is `a` `b` `c`
In final mode, the input is partial data **and accumulated to "xx n"**

If you have this data as input:

```
a 1
b 2
c 2
```

In partial mode, the output is

```
a 1
b 1
c 1
```

In final mode, the output is

```
a 1
b 2
c 2
```


Back to the topic, we have push down agg to partition unions.
If the push-downed agg is not made to partial, 

```
create table t (c1 char, c2 int) partition by hash(c1) partitions 3;
insert into t values ('a', 1),('b', 2),('c', 3),('b', 4),('c', 5);
select count(*) from t group by c1;
```

We get wrong result:

```
a 1                   a 1                  a 1
b 2                   b 6                  b 6 
c 3  (final agg) =>   c 8  (final agg) =>  c 8
b 4
c 5
```

The correct result should be:

```
a 1                     a 1                  a 1
b 2                     b 2                  b 2
c 3  (partial agg) =>   c 2  (final agg) =>  c 2
b 4
c 5
```

Comment (tiancaiamao) [reply planner/core/rule_aggregation_push_down.go:442]: @bb7133

**Pull request merged**